pkg/arvo/tests/sys/zuse/crypto/aes.hoon

```diff
@@ -0,0 +1,109 @@
+::  tests for the aes block cipher
+::
+::  test vectors from Appendix C of:
+::  https://csrc.nist.gov/publications/fips/fips197/fips-197.pdf
+::
+/+  *test
+=,  aes:crypto
+::
+|%
+::
+++  do-test-vectors-ecba
+  |=  ves=(list [key=@ux in=@ux out=@ux])
+  ^-  tang
+  ?~  ves  ~
+  =-  (weld - $(ves t.ves))
+  =,  i.ves
+  ;:  weld
+    ::  encrypt
+    %+  expect-eq
+        !>  out
+        !>  `@ux`(~(en ecba key) in)
+    ::  decrypt
+    %+  expect-eq
+        !>  in
+        !>  `@ux`(~(de ecba key) out)
```

Comment:
This feels like it could be made much DRYer -- the `do-test-vectors-ecbX` functions are identical except for whether they use `ecba`, `ecbb`, or `ecbc`. Is there a way to accept the algorithm as an argument?

Reply:
Good question! Perhaps this?

```hoon
+$  vector  [key=@ux in=@ux out=@ux]
::
++  run-vectors-on
  |*  [ecbx=_ecba vectors=(list vector)]
  ^-  tang
  ::  rest of the logic. maybe use +turn in place of manual looping?
  ::  example encryption call:
  (~(en ecbx key) in)
::
++  test-vectors-ecba
  %+  run-vectors-on  ecba
  :~  some
      vectors
  ==
```

This depends on the fact that all the `+ecbx:aes:crypto` doors have the same shape/arms (as `+ecba`), but I think that's a sane & safe assumption.

Reply:
ah, that's clever! Tested locally and it seems to work

Reply:
Can you elaborate on how to use `turn` here? I can sort of picture how it would work (map the encrypt and decrypt checks over each vector), but I can't get it to work. Here's what I have:
```hoon
++  do-test-vectors
  |*  [ecbx=_ecba ves=(list vector)]
  ^-  tang
  =<  (weld (turn ves enc) (turn ves dec))
  |%
  ++  enc
    |=  v=vector
    %+  expect-eq
      !>  out.v
      !>  `@ux`(~(en ecbx key.v) in.v)
  ++  dec
    |=  v=vector
    %+  expect-eq
      !>  in.v
      !>  `@ux`(~(de ecbx key.v) out.v)
  --
```

Reply:
That seems pretty close. The only thing you're really missing is a `+zing` to turn the resulting `(list tang)` into a `tang`. (`tang` is a `(list tank)`, and `+zing` turns a list of lists into a single list.)

With `|^` as shorthand for `=<  ...  |%`, and using `+category` from `/lib/test` for fancier failure printing, I'd do something like this:

```hoon
++  do-test-vectors
  |*  [ecbx=_ecba ves=(list vector)]
  ^-  tang
  |^  %+  weld
        %+  category  "encrypting"
        (zing (turn ves enc))
      %+  category  "decrypting"
      (zing (turn ves dec)))
  ::
  ++  enc
    |=  vector
    %+  expect-eq
      !>  out
      !>  `@ux`(~(en ecbx key) in)
  ::
  ++  dec
    |=  vector
    %+  expect-eq
      !>  in
      !>  `@ux`(~(de ecbx key) out)
  --
```

(If that nest-fails around the `zing`s, it may need a cast to the expected type.)